runtime/event_loop.cpp

```diff
@@ -38,38 +39,51 @@ bool EventLoop::cancel_async_task(api::Engine *engine, const int32_t id) {
 
 bool EventLoop::has_pending_async_tasks() { return !queue.get().tasks.empty(); }
 
-bool EventLoop::run_event_loop(api::Engine *engine, double total_compute) {
-  // Loop until no more resolved promises or backend requests are pending.
-  // LOG("Start processing async jobs ...\n");
+void EventLoop::incr_event_loop_lifetime() {
+  queue.get().lifetime_cnt++;
+}
+
+void EventLoop::decr_event_loop_lifetime() {
+  MOZ_ASSERT(queue.get().lifetime_cnt > 0);
+  queue.get().lifetime_cnt--;
+}
+
+bool lifetime_complete() {
+  return queue.get().lifetime_cnt == 0;
+}
 
+bool EventLoop::run_event_loop(api::Engine *engine, double total_compute) {
   JSContext *cx = engine->cx();
 
-  do {
-    // First, drain the promise reactions queue.
+  while (true) {
     while (js::HasJobsPending(cx)) {
       js::RunJobs(cx);
-
       if (JS_IsExceptionPending(cx))
         return false;
+      if (lifetime_complete())
+        return true;
     }
 
-    // TODO: add general mechanism for extending the event loop duration.
-    // Then, check if the fetch event is still active, i.e. had pending promises
-    // added to it using `respondWith` or `waitUntil`.
-    // if (!builtins::FetchEvent::is_active(fetch_event))
-    //   break;
+    auto tasks = &queue.get().tasks;
 
-    // Process async tasks.
-    if (has_pending_async_tasks()) {
-      auto tasks = &queue.get().tasks;
-      const auto index = api::AsyncTask::select(tasks);
+    // Unresolved lifetime error -
+    // if there are no async tasks, and the lifetime was not complete
+    // then we cannot complete the lifetime
+    if (tasks->size() == 0) {
+      return false;
+    }
+
+    const auto ready = api::AsyncTask::poll(tasks);
+    for (auto index : ready) {
       auto task = tasks->at(index);
       if (!task->run(engine)) {
         return false;
       }
       tasks->erase(tasks->begin() + index);
+      if (lifetime_complete())
+        return true;
     }
-  } while (js::HasJobsPending(engine->cx()) || has_pending_async_tasks());
+  }
```

Comment:
Thanks for the point about the erasure. To address your points:

1. I've now added some code that checks that the ready list is sorted, and if it is not will first sort it, before running through the tasks. Task ordering should imply completion ordering so this seems fine to assume. Then with this monotonic assumption, we can handle the ready index by just decrementing the index by the number of erasures in the loop fixing this erasure offset issue.
2. I'm not sure we want to do a drain per task here for maximum parallelism. `task->run()` will do some immediate synchronous work, which in turn will have its own task queues and follow-up work. If we just drain directly all those sync task queues, we don't necessarily maximally interleave parallel work, which seems desireable in JS environments. Therefore I think the code as written is actually correct from a task management perspective, since the immediate reactions then properly interleave with other parallel work. For now, I've added a single `RunJobs` call along with this so that each individual task will get its immediate processing, followed by its immediate reactions, but my preference would still be to leave `RunJobs` out of here entirely. Well-written async code in JS should have stall points such that maximal interleaving is desired.
3. Another case to guard against is event loop reentrancy in this model. I've added a guard to ensure there is no reentrancy now. If you can think of any other edge cases please let me know.